doc_internal/rust-devel.md

```diff
@@ -0,0 +1,162 @@
+
+fish-shell currently depends on Rust 1.67 or later. To install Rust, follow https://rustup.rs.
+
+### Build via CMake
+
+It is recommended to build inside `fish-shell/build`. This will make it easier for Rust to find the `config.h` file.
+
+Build via CMake as normal (use any generator, here we use Ninja):
+
+```shell
+$ cd fish-shell
+$ mkdir build && cd build
+$ cmake -G Ninja ..
+$ ninja
+```
+
+This will create the usual fish executables.
+
+### Build just libfish_rust.a with Cargo
+
+The directory `fish-rust` contains the Rust sources. These require that CMake has been run to produce `config.h` which is necessary for autocxx to succeed.
+
+Follow the "Build from CMake" steps above, and then:
+
+```shell
+$ cd fish-shell/fish-rust
+$ cargo build
+```
+
+This will build only the library, not a full working fish, but it allows faster iteration for Rust development. That is, after running `cmake` you can open the `fish-rust` as the root of a Rust crate, and tools like rust-analyzer will work.
+
+## Development
+
+The basic development loop for this port:
+
+1. Pick a .cpp (or in some cases .h) file to port, say `util.cpp`.
+2. Add the corresponding `util.rs` file to `fish-rust/`.
+3. Reimplement it in Rust, along with its dependencies as needed. Match the existing C++ code where practical, including propagating any relevant comments.
+   - Do this even if it results in less idiomatic Rust, but avoid being super-dogmatic either way.
+   - One technique is to paste the C++ into the Rust code, commented out, and go line by line.
+4. Decide whether any existing C++ callers should invoke the Rust implementation, or whether we should keep the C++ one.
+   - Utility functions may have both a Rust and C++ implementation. An example is `FLOG` where interop is too hard.
+   - Major components (e.g. builtin implementations) should _not_ be duplicated; instead the Rust should call C++ or vice-versa.
+
+You will likely run into limitations of [`autocxx`](https://google.github.io/autocxx/) and to a lesser extent [`cxx`](https://cxx.rs/). See the FFI sections below.
+
+## Type Mapping
+
+### Strings
+
+Fish will mostly _not_ use Rust's `String/&str` types as these cannot represent non-UTF8 data using the default encoding.
+
+fish's primary string types will come from the [`widestring` crate](https://docs.rs/widestring). The two main string types are `WString` and `&wstr`, which are renamed [Utf32String](https://docs.rs/widestring/latest/widestring/utfstring/struct.Utf32String.html) and [Utf32Str](https://docs.rs/widestring/latest/widestring/utfstr/struct.Utf32Str.html). `WString` is an owned, heap-allocated UTF32 string, `&wstr` a borrowed UTF32 slice.
+
+In general, follow this mapping when porting from C++:
+
+- `wcstring` -> `WString`
+- `const wcstring &` -> `&wstr`
+- `const wchar_t *` -> `&wstr`
+
+None of the Rust string types are nul-terminated. We're taking this opportunity to drop the nul-terminated aspect of wide string handling.
```

Comment:
Typo? 
```suggestion
None of the Rust string types are null-terminated. We're taking this opportunity to drop the nul-terminated aspect of wide string handling.
```

Reply:
Nope - ascii calls it "NUL"

Reply:
From [Wikipedia: Null character](https://en.wikipedia.org/w/index.php?title=Null_character&oldid=1123297106):
> It is often abbreviated as **NUL** (or **NULL**, though in some contexts that term is used for the null pointer).

Reply:
This is bikeshedding; I am hiding this thread.

Reply:
> Nope - [ascii calls it "NUL"](https://datatracker.ietf.org/doc/html/rfc20#section-5.2)

ugh, dunning-kruger hits way too hard on me. thanks for the info